contracts/Proposal.sol

```diff
@@ -0,0 +1,120 @@
+ *
+ * This contract is intended to be used in one of two possible ways. Either:
+ * - A target RSV basket is suggested, and quantities to be exchanged are  
+ *     deduced at the time of proposal execution.
+ * - A specific quantity of tokens to be exchanged is suggested, and the 
+ *     resultant RSV basket is determined at the time of proposal execution.
+ *
+ */
+contract Proposal is Ownable {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    uint256 public id;
+    uint256 public time;
+    address public proposer;
+    address[] public tokens;
+    uint256[] public quantitiesIn; // token quantities to be added to the Vault
+    uint256[] public quantitiesOut; // total quantities to be withdrawn from the Vault
+
+    enum Statuses { Created, Accepted, Closed, Completed }
+    Statuses status;
+
+    uint256[] basketBacking;
+    Basket public basket;
+
+
+    constructor(
+        uint256 _id,
+        address _proposer,
+        address[] memory _tokens,
+        uint256[] memory _quantitiesIn, // total quantities for the entire RSV supply, not per RSV
+        uint256[] memory _quantitiesOut, // total quantities for the entire RSV supply, not per RSV
+        Basket _basket
+    ) 
+        public 
+    {
+        require(_quantitiesIn.length == _quantitiesOut.length, "quantities mismatched");
+        (id, proposer, tokens, quantitiesIn, quantitiesOut, basket) = 
+            (_id, _proposer, _tokens, _quantitiesIn, _quantitiesOut, _basket);
+        status = Statuses.Created;
+    }
+
+    /// Helper to read status from off-chain. 
+    function getStatus() external view returns (Statuses) {
+        return status;
+    }
+
+    /// Moves a proposal from the Created to Accepted state. 
+    function accept(uint256 _time) external onlyOwner {
+        require(status == Statuses.Created, "proposal not created");
+        time = _time;
+        status = Statuses.Accepted;
+    }
+
+    /// Closes a proposal if it has not been completed. 
+    function close() external onlyOwner {
+        require(status != Statuses.Completed);
+        status = Statuses.Closed;
+    }
+
+    /// Moves a proposal from the Accepted to Completed state. 
+    /// Returns the tokens, quantitiesIn, and quantitiesOut, required to implement the proposal.
+    function complete(
+        uint256 _rsvSupply, 
+        address _vaultAddr, 
+        Basket _prevBasket
+    ) 
+        external onlyOwner returns(address[] memory, uint256[] memory, uint256[] memory) 
+    {
+        require(status == Statuses.Accepted, "proposal must be accepted");
+        require(now > time, "wait to execute");
+
+        if (basket == Basket(0)) {
+            uint256 newQuantity;
+            for (uint i = 0; i < tokens.length; i++) {
+                newQuantity = IERC20(tokens[i]).balanceOf(_vaultAddr) + quantitiesIn[i] - quantitiesOut[i];
+                require(newQuantity >= 0, "proposal removes too many tokens");
```

Comment:
It seems to me like this might prevent changes that remove a token type from the basket, unless you leave one of them in the backing at the end.

That may not be true -- there might be another way to do that -- but it also seems like a bit of a weird check, given that it will only catch reducing the supply to exactly zero (it can't catch underflows b/c this is an unsigned int).

Reply:
Caught in refactoring. See `thursday`, or later branches.